codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2PushPromiseFrameTest.java

```diff
@@ -0,0 +1,233 @@
+
+                        Http2FrameCodec frameCodec = Http2FrameCodecBuilder.forServer()
+                                .autoAckSettingsFrame(true)
+                                .autoAckPingFrame(true)
+                                .build();
+
+                        pipeline.addLast(frameCodec);
+                        pipeline.addLast(new ServerHandler());
+                    }
+                });
+
+        ChannelFuture channelFuture = serverBootstrap.bind(0).sync();
+
+        final Bootstrap bootstrap = new Bootstrap()
+                .group(eventLoopGroup)
+                .channel(NioSocketChannel.class)
+                .handler(new ChannelInitializer<SocketChannel>() {
+                    @Override
+                    protected void initChannel(SocketChannel ch) {
+                        ChannelPipeline pipeline = ch.pipeline();
+
+                        Http2FrameCodec frameCodec = Http2FrameCodecBuilder.forClient()
+                                .autoAckSettingsFrame(true)
+                                .autoAckPingFrame(true)
+                                .initialSettings(Http2Settings.defaultSettings().pushEnabled(true))
+                                .build();
+
+                        pipeline.addLast(frameCodec);
+                        pipeline.addLast(clientHandler);
+                    }
+                });
+
+        connectionFuture = bootstrap.connect(channelFuture.channel().localAddress());
+    }
+
+    @Test
+    public void send() {
+        connectionFuture.addListener(new ChannelFutureListener() {
+            @Override
+            public void operationComplete(ChannelFuture future) {
+                clientHandler.write();
+            }
+        });
+    }
+
+    @After
+    public void shutdown() {
+        eventLoopGroup.shutdownGracefully();
+    }
+
+    private final class ServerHandler extends Http2ChannelDuplexHandler {
+
+        @Override
+        public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+
+            if (msg instanceof Http2HeadersFrame) {
+                Http2HeadersFrame receivedFrame = (Http2HeadersFrame) msg;
+
+                Http2Headers pushRequestHeaders = new DefaultHttp2Headers();
+                pushRequestHeaders.path("/meow")
+                        .method("GET")
+                        .scheme("https")
+                        .authority("localhost:5555");
+
+                // Write PUSH_PROMISE request headers
+                Http2FrameStream newPushFrameStream = newStream();
+                Http2PushPromiseFrame pushPromiseFrame = new DefaultHttp2PushPromiseFrame(pushRequestHeaders);
+                pushPromiseFrame.stream(receivedFrame.stream());
+                pushPromiseFrame.pushStream(newPushFrameStream);
+                ctx.writeAndFlush(pushPromiseFrame);
+
+                contentMap.put(newPushFrameStream.id(), "Meow, I am Pushed via HTTP/2");
```

Comment:
Honestly I think this should be done in a `ChannelFutureListener` so you are sure the write was actually done and so the id was set... Basically move this line and everything below in a a `ChannelFutureListener` that is added to the headers write.